drake/examples/HolonomicDrive/HolonomicDrivePlant.m

```diff
@@ -0,0 +1,246 @@
+        i = i + 1;
+      end
+    end
+
+    function [qdd, qdddx] = sodynamics(obj,t,q,qd,u)
+      n = length(obj.wheels);
+
+      if nargout > 1
+        qdddx = zeros(3, 7 + n);
+        % dynamics are linear in each wheel.
+        for i = 1:n
+          wheel = obj.wheels(i);
+          qdddxi = oneWheelGradients(obj.m, obj.I, wheel, t, q, qd, u(i));
+          qdddx(:,1:7) = qdddx(:,1:7) + qdddxi(:,1:7);
+
+          % qdd / du_i
+          qdddx(:,7+i) = qdddxi(:,8);
+        end
+      end
+
+      theta = q(3);
+      rotation = [[cos(theta); sin(theta)] [-sin(theta); cos(theta)]];
+
+      bodyvel = rotation' * qd(1:2);
+
+      speeds = obj.rotorSpeeds(bodyvel, qd(3));
+
+
+      total_force = [0; 0];
+      total_moment = 0;
+      for i = 1:n
+        wheel = obj.wheels(i);
+        % simplified motor dynamics
+        tau = wheel.tauMax * (u(i) - speeds(i) / wheel.omegaMax);
+
+        % assign direction to force
+        forceDir = [-wheel.slipDir(2); wheel.slipDir(1)];
+        f = tau / wheel.r * forceDir / (forceDir' * wheel.driveDir);
+
+        % combine forces
+        total_force = total_force + f;
+        total_moment = total_moment...
+                   + wheel.pos(1) * f(2)...
+                   - wheel.pos(2) * f(1);
+      end
+
+      % convert force to world space
+      total_force = rotation * total_force;
+
+      qdd = [total_force / obj.m; total_moment / obj.I];
+    end
+
+    function [utraj,xtraj] = optimalTrajectory(p,x0,xf, tf0)
+      N = 51;
+      prog = DircolTrajectoryOptimization(p,N,[tf0/2 tf0*2]);
+      prog = prog.addInputConstraint(BoundingBoxConstraint(p.umin, p.umax), 1:N);
+      prog = prog.addStateConstraint(ConstantConstraint(x0), 1);
+      prog = prog.addStateConstraint(ConstantConstraint(xf), N);
+      prog = prog.addFinalCost(@finalCost);
+
+      function [h,dh] = finalCost(t,x)
+        h = t;
+        dh = [1,zeros(1,size(x,1))];
+      end
+
+      traj_init.x = PPTrajectory(foh([0,tf0],[x0,xf]));
+      info = 0;
+      while info~=1
+        tic
+        [xtraj,utraj,z,F,info] = prog.solveTraj(tf0, traj_init);
+        toc
+      end
+    end
+
+    function v = constructVisualizer(obj)
+      v = HolonomicDriveVisualizer(obj);
+    end
+
+    function overall_traj = appendWheelStates(obj, xtraj)
+      % given a trajectory, augments it with wheel angle information, useful for rendering
+      % Only a convenience method - use HolonomicDriveWheelSensors directly for when this
+      % information is used numerically.
+      sensors = HolonomicDriveWheelSensors(obj);
+      wheeltraj = simulate(cascade(xtraj, sensors), xtraj.tspan, zeros(length(obj.wheels), 1));
+      overall_traj = [xtraj; wheeltraj];
+    end
+  end
+
+  methods(Static)
+    function obj = unitModel(n)
+      % Construct a drive where all params are equal to 1, just like
+      % in the real world. n is the number of wheels, and defaults to 3
+      r = 1;
+      if ~exist('n', 'var')

```

Comment:
Hmm, I like this; we should have used this everywhere instead of doing nargin checks.